odl/solvers/functional/example_functionals.py

```diff
@@ -0,0 +1,149 @@
+
+    """The famous Rosenbrock funciton on ``R^n``.
+
+    Notes
+    -----
+    The functional is defined for :math:`x` in :math:`\\mathbb{R}^n`,
+    :math:`n \\geq 2`, as
+
+    .. math::
+        \sum_{i=1}^{n - 1} c (x_{i+1} - x_i^2)^2 + (1 - x_i)^2
+
+    Where :math:`c` is a constant usually set to 100 which determines how ill-
+    posed the function should be.
+    It has a minimum at :math:`x = [1, \\dots, 1]`, independent of :math:`c`.
+    """
+
+    def __init__(self, space, scale=100.0):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `FnBase`
+            Domain of the functional.
+        scale : positive float, optional
+            The scale ``c`` in the functional, determines how ill posed the
+            functional should be.
+
+        Examples
+        --------
+        >>> import odl
+        >>> r2 = odl.rn(2)
+        >>> functional = RosenbrockFunctional(r2)
+        >>> functional([1, 1])  # optimum is 0 at [0, 0]
+        0.0
+        >>> functional([0, 1])
+        200.0
+
+        Can also be used in higher dimensions
+
+        >>> r5 = odl.rn(5)
+        >>> functional = RosenbrockFunctional(r5)
+        >>> functional([1, 1, 1, 1, 1])
+        0.0
+        """
+        self.scale = float(scale)
+        if not isinstance(space, FnBase):
+            raise ValueError('`space` must be a `FnBase`')
+        if space.size < 2:
+            raise ValueError('`space` must be at least two dimensional')
+        super().__init__(space=space, linear=False, grad_lipschitz=np.nan)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        result = 0
+        for i in range(0, self.domain.size - 1):
+            result += self.scale * ((x[i + 1] - x[i] ** 2) ** 2 +
+                                    (x[i] - 1) ** 2)
+        return result
+
+    @property
+    def gradient(self):
+        """Gradient operator of the Rosenbrock functional."""
+        functional = self
+        c = self.scale
+
+        class RosenbrockGradient(Operator):
+
+            """The gradient operator of the Rosenbrock functional."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x, out):
+                """Apply the gradient operator to the given point."""
+                for i in range(1, self.domain.size - 1):
+                    out[i] = (2 * c * (x[i] - x[i - 1]**2) -
+                              4 * c * (x[i + 1] - x[i]**2) * x[i] -
+                              2 * (1 - x[i]))
+                out[0] = (-4 * c * (x[1] - x[0] ** 2) * x[0] +
+                          2 * (x[0] - 1))
+                out[-1] = 2 * c * (x[-1] - x[-2] ** 2)
+
+            def derivative(self, x):
+                """The derivative of the gradient.
+
+                This is also known as the hessian.
```

Comment:
Even worse than the gradient, for these things I think it's better to just let users read the code (it is readily available) if they are interested. Easier to maintain as well.

Adding it would only be helpfull (imo) if we also added a deriviation of the result, which would take way too much space.